include/clangmetatool/include_graph_dependencies.h

```diff
@@ -45,7 +46,39 @@ struct IncludeGraphDependencies {
    */
   static std::set<clangmetatool::types::FileUID>
   liveDependencies(const clangmetatool::collectors::IncludeGraphData *data,
-                   const clangmetatool::types::FileUID &headerFUID);
+                   const clangmetatool::types::FileUID &fileFUID);
+
+  /*
+   * A data structure for include graph weak dependencies analysis
+   * for a specific source file
+   *
+   * Key: all headers the file directly depends on
+   * Value: a set of direct included headers that could allow the file to
+   * access the key header
+   *
+   * Example:
+   * { "def1.h": {"a.h", "b.h"},
+   *   "def2.h": {"a.h", "c.h"} }
+   *
+   * So that given the example data above it means current analyzing file:
+   * - depends on definitions from "def1.h", "def2.h" directly
+   * - includes "a.h", "b.h", "c.h"
+   * - can access "def1.h" by "a.h" and "b.h"
+   * - can access "def2.h" by "a.h" and "c.h"
+   */
+  typedef std::map<clangmetatool::types::FileUID,
+                   std::set<clangmetatool::types::FileUID>> DirectDependenciesMap;
+
+  /**
+   * Get the 'live' weak dependencies of a header within the given include graph.
```

Comment:
I don't think this documentation is clear on what "weak" means or how `liveWeakDependencies` differs from `liveDependencies`. The idea is that the old function shows the first directly included header that leads to each "live dependency", while the latter shows all headers that could lead to some "live dependency" (which are potentially "weak" because you only need one of them) sorted by the header that declares that dependency? Is this accurate?

Maybe dropping the word weak would be clearer, and calling this something like `findAllPathsToLiveDependencies`? `liveDependencyPaths`?

Reply:
I agree, in that the `weak` part meant `a.h` & `b.h` are weak dependencies because either of them could be removed & the code would still work.

But it doesn't find paths as well. It finds all the indirect dependencies & the header in the source that can lead to them.

`findDependencyProviders` seems closer to what it does, but I don't really like it because its incredibly vague.

Reply:
Hrm, looking again at the definition above of "live dependency" it seems like it refers to the directly included header as a "live dependency", not the final header that has the declaration needed. That makes my description above incorrect. So a "live dependency" is "weak" when the final header that is accessible through it is also potentially accessible through other direct includes?

I guess the function name is fine as long as this is described precisely, e.g. unlike `liveDependencies`, which returns only the first header that leads to a header with a needed declaration, the output of this function includes all direct includes that have a path to a header with a needed declaration.